app/src/main/java/com/cyberbot/checkers/game/Grid.java

```diff
@@ -80,4 +118,67 @@ public boolean attemptMove(GridEntry src, GridEntry dst) {
     public Iterator<GridEntry> iterator() {
         return gridEntries.iterator();
     }
+
+    private HashSet<GridEntry> getAdjacentEntries(@NotNull GridEntry entry) {
+        HashSet<GridEntry> adjacentEntries = new HashSet<>();
+
+        // Short name for code readability
+        final int x = entry.getX();
+        final int y = entry.getY();
+
+        // Bottom corner
+        if(x == 0 && y == size-1) {
+            adjacentEntries.add(getEntryByCoords(1, size-2));
+        }
+        // Upper corner
+        else if(y == 0 && x == size-1) {
+            adjacentEntries.add(getEntryByCoords(size-2, 1));
+        }
+        // Upper edge
+        else if(x == 0) {
+            adjacentEntries.add(getEntryByCoords(1, y-1));
+            adjacentEntries.add(getEntryByCoords(1, y+1));
+        }
+        // Left edge
+        else if(y == 0) {
+            adjacentEntries.add(getEntryByCoords(x-1, 1));
+            adjacentEntries.add(getEntryByCoords(x+1, 1));
+        }
+        // Bottom edge
+        else if(x == size-1) {
+            adjacentEntries.add(getEntryByCoords(x-1, y-1));
+            adjacentEntries.add(getEntryByCoords(x-1, y+1));
+        }
+        // Right edge
+        else if(y == size-1) {
+            adjacentEntries.add(getEntryByCoords(x-1, y-1));
+            adjacentEntries.add(getEntryByCoords(x+1, y-1));
+        }
+        // Mid entries
+        else {
+            adjacentEntries.add(getEntryByCoords(x-1, y-1));
+            adjacentEntries.add(getEntryByCoords(x-1, y+1));
+            adjacentEntries.add(getEntryByCoords(x+1, y-1));
+            adjacentEntries.add(getEntryByCoords(x+1, y+1));
+        }
+
+        return adjacentEntries;
+    }
+
+    public HashSet<GridEntry> calculateAllowedMoves(@NotNull GridEntry entry, boolean storeInCache) {
+        HashSet<GridEntry> allowedMoves = new HashSet<>();
+        for(GridEntry adjEntry: getAdjacentEntries(entry)) {
+            if(adjEntry.getPlayer() == PlayerNum.NOPLAYER) {
+                if(canMoveBackwards
+                        || (entry.getPlayer() == PlayerNum.FIRST && entry.getY() < adjEntry.getY())
+                        || (entry.getPlayer() == PlayerNum.SECOND && entry.getY() > adjEntry.getY())) {
+                    allowedMoves.add(adjEntry);
+                }
+            }
+        }
```

Comment:
For now its fine for the method to be `public`, however later I'd suggest creating a `public getAllowedMoves(GridEntry)`, which first checks whether the moves are cached and only then calls this method, which should be made `private` by then